### PR TITLE
Issue 625 Pipeline run pause/restart scheduling

### PR DESCRIPTION
**Pull request by Wedds**
This PR solves the issue #625.

It brings pause and restarts scheduling for a run, described by cron expression. The user is able to specify any number of rules, but it is forbidden to set conflicting rules (i.e. with identical cron expressions), as well as pause scheduling for cluster runs.

Changes summary:
- Create an extra table in DB for storing runs' scheduling rules
- Add new functionality to PipelineRun components, create entities for runs' scheduling information storing, implement new services: PipelineRunScheduler, PipelineRunScheduleManager
- Introduce API for CRUD operations with runs' schedules in PipelineRunController

Execution of scheduled tasks is performed using [Quartz](http://www.quartz-scheduler.org/).

### CODE REVIEW

File: api/src/main/java/com/epam/pipeline/acl/run/RunApiService.java [hunk 1]
```diff
@@ -267,4 +271,33 @@ public Boolean checkFreeSpaceAvailable(final Long runId) {
     public PipelineRun terminateRun(final Long runId) {
         return runManager.terminateRun(runId);
     }
+
+    @PreAuthorize(RUN_ID_EXECUTE)
```
Comment (mzueva) [inline api/src/main/java/com/epam/pipeline/acl/run/RunApiService.java]: I think it is better to have a separate `RunScheduleApiService`

File: api/src/main/java/com/epam/pipeline/controller/pipeline/PipelineRunController.java [hunk 2]
```diff
@@ -486,4 +490,62 @@ public Result prolongIdleRun(@PathVariable(value = RUN_ID) Long runId) {
                                               @RequestBody final TagsVO tagsVO) {
         return Result.success(runApiService.updateTags(runId, tagsVO));
     }
+
+    @PostMapping(value = "/run/{runId}/schedule/create")
```
Comment (mzueva) [inline api/src/main/java/com/epam/pipeline/controller/pipeline/PipelineRunController.java]: Same as above - add a separate controller

File: api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java [hunk 3]
```diff
@@ -345,7 +345,7 @@ private void performNotify(PipelineRun run, double cpuUsageRate,
     }
 
     private void performStop(PipelineRun run, double cpuUsageRate) {
-        if (run.isNonPause() || isClusterRun(run)) {
+        if (run.isNonPause() || PipelineRunManager.isClusterRun(run)) {
```
Comment (mzueva) [inline api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java]: This method was changed in `develop` branch - please rebase

File: api/src/main/java/com/epam/pipeline/manager/pipeline/PipelineRunScheduleManager.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+@Service
+@Slf4j
+public class PipelineRunScheduleManager {
+
+    private final PipelineRunScheduleDao runScheduleDao;
+    private final PipelineRunManager pipelineRunManager;
+    private final PipelineRunScheduler scheduler;
+    private final MessageHelper messageHelper;
+
+    @Autowired
```
Comment (mzueva) [inline api/src/main/java/com/epam/pipeline/manager/pipeline/PipelineRunScheduleManager.java]: You can use `@RequiredArgsConstructor`

**Review by mzueva — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/epam/pipeline/manager/pipeline/PipelineRunScheduleManager.java (hunk 4 above)
Comment (Wedds) [reply api/src/main/java/com/epam/pipeline/manager/pipeline/PipelineRunScheduleManager.java]: Done

File: api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java (hunk 3 above)
Comment (Wedds) [reply api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java]: Done

File: api/src/main/java/com/epam/pipeline/controller/pipeline/PipelineRunController.java (hunk 2 above)
Comment (Wedds) [reply api/src/main/java/com/epam/pipeline/controller/pipeline/PipelineRunController.java]: Done

File: api/src/main/java/com/epam/pipeline/acl/run/RunApiService.java (hunk 1 above)
Comment (Wedds) [reply api/src/main/java/com/epam/pipeline/acl/run/RunApiService.java]: Done

**Review by mzueva — CHANGES_REQUESTED** [review]
Could you please perform checkstyle & pmd fix and remove `PipelineRunSchedulerPureJava` class.

**Pull request merged**